policy/policy-data.go

```diff
@@ -0,0 +1,141 @@
+// Copyright 2015 ISRG.  All rights reserved
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package policy
+
+import (
+	"fmt"
+	"strings"
+
+	blog "github.com/letsencrypt/boulder/log"
+	"github.com/letsencrypt/boulder/sa"
+
+	gorp "github.com/letsencrypt/boulder/Godeps/_workspace/src/gopkg.in/gorp.v1"
+)
+
+const whitelisted = "whitelist"
+const blacklisted = "blacklist"
+
+// DomainRule ...
+type DomainRule struct {
+	Host string `db:"host"`
+	Type string `db:"type"`
+}
+
+func reverseName(domain string) string {
+	labels := strings.Split(domain, ".")
+	for i, j := 0, len(labels)-1; i < j; i, j = i+1, j-1 {
+		labels[i], labels[j] = labels[j], labels[i]
+	}
+	return strings.Join(labels, ".")
+}
+
+// PolicyAuthorityDatabaseImpl enforces policy decisions based on various rule
+// lists
+type PolicyAuthorityDatabaseImpl struct {
+	log   *blog.AuditLogger
+	dbMap *gorp.DbMap
+}
+
+// NewPolicyAuthorityDatabaseImpl constructs a Policy Authority Database (and
+// creates tables if they are non-existent)
+func NewPolicyAuthorityDatabaseImpl(name string) (padb *PolicyAuthorityDatabaseImpl, err error) {
+	logger := blog.GetAuditLogger()
+	dbMap, err := sa.NewDbMap(name)
+	if err != nil {
+		return nil, err
+	}
+
+	dbMap.AddTableWithName(DomainRule{}, "ruleList").SetKeys(false, "Host")
+
+	err = dbMap.CreateTablesIfNotExists()
+	if err != nil {
+		return
+	}
+
+	padb = &PolicyAuthorityDatabaseImpl{
+		dbMap: dbMap,
+		log:   logger,
+	}
+
+	return padb, nil
+}
+
+// LoadRules loads the whitelist and blacklist into the database in a transaction
+// deleting any previous content
+func (padb *PolicyAuthorityDatabaseImpl) LoadRules(rules []DomainRule) error {
+	tx, err := padb.dbMap.Begin()
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+	_, err = tx.Exec("DELETE FROM ruleList")
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+
+	for _, r := range rules {
+		r.Host = reverseName(r.Host)
+		tx.Insert(&r)
+	}
+
+	err = tx.Commit()
+	return err
+}
+
+// DumpRules retrieves all DomainRules in the database so they can be written to
+// disk
+func (padb *PolicyAuthorityDatabaseImpl) DumpRules() ([]DomainRule, error) {
+	var dR []DomainRule
+	_, err := padb.dbMap.Select(&dR, "SELECT * FROM ruleList")
+	for _, r := range dR {
+		r.Host = reverseName(r.Host)
+	}
+
+	return dR, err
+}
+
+// CheckRules will query the database for white/blacklist rules that match host,
+// if both whitelist and blacklist rules are found the whitelist will always win
+func (padb *PolicyAuthorityDatabaseImpl) CheckRules(host string, requireWhitelisted bool) error {
+	host = reverseName(host)
+	var rules []DomainRule
+	_, err := padb.dbMap.Select(
+		&rules,
+		`SELECT * FROM ruleList WHERE :host >= host ORDER BY host ASC`,
```

Comment:
In order for our order query to work correctly, we need to enforce an invariant in LoadRules: No name on the blacklist can be a suffix of another name. For instance, we can't have both `foo.example.com` and `example.com`. In such a situation `foo.example.com` would be redundant anyhow. You can achieve this invariant by sorting all the rules and iterating over once, checking that no reversed name is a prefix of it's next neighbor.

Additionally, proper semantics require that we query separately for the whitelist and blacklist, both for order reasons and because we want to clearly specify which takes precedence (the blacklist).

When implementing the above, please make sure to document the reasoning in comments.